src/main/java/com/puppycrawl/tools/checkstyle/SarifLogger.java

```diff
@@ -0,0 +1,348 @@
+import com.puppycrawl.tools.checkstyle.api.SeverityLevel;
+
+/**
+ * Simple SARIF logger.
+ * SARIF stands for the static analysis results interchange format.
+ * Reference: https://sarifweb.azurewebsites.net/
+ */
+public class SarifLogger extends AutomaticBean implements AuditListener {
+
+    /** The length of unicode placeholder. */
+    private static final int UNICODE_LENGTH = 4;
+
+    /** Unicode escaping upper limit. */
+    private static final int UNICODE_ESCAPE_UPPER_LIMIT = 0x1F;
+
+    /** Input stream buffer size. */
+    private static final int BUFFER_SIZE = 1024;
+
+    /** The placeholder for message. */
+    private static final String MESSAGE_PLACEHOLDER = "${message}";
+
+    /** The placeholder for severity level. */
+    private static final String SEVERITY_LEVEL_PLACEHOLDER = "${severityLevel}";
+
+    /** The placeholder for uri. */
+    private static final String URI_PLACEHOLDER = "${uri}";
+
+    /** The placeholder for line. */
+    private static final String LINE_PLACEHOLDER = "${line}";
+
+    /** The placeholder for column. */
+    private static final String COLUMN_PLACEHOLDER = "${column}";
+
+    /** The placeholder for rule id. */
+    private static final String RULE_ID_PLACEHOLDER = "${ruleId}";
+
+    /** The placeholder for version. */
+    private static final String VERSION_PLACEHOLDER = "${version}";
+
+    /** The placeholder for results. */
+    private static final String RESULTS_PLACEHOLDER = "${results}";
+
+    /** Helper writer that allows easy encoding and printing. */
+    private final PrintWriter writer;
+
+    /** Close output stream in auditFinished. */
+    private final boolean closeStream;
+
+    /** The results. */
+    private final List<String> results = new ArrayList<>();
+
+    /** Content for the entire report. */
+    private final String report;
+
+    /** Content for result representing an error with source line and column. */
+    private final String resultLineColumn;
+
+    /** Content for result representing an error with source line only. */
+    private final String resultLineOnly;
+
+    /** Content for result representing an error with filename only and without source location. */
+    private final String resultFileOnly;
+
+    /** Content for result representing an error without filename or location. */
+    private final String resultErrorOnly;
+
+    /**
+     * Creates a new {@code SarifLogger} instance.
+     *
+     * @param outputStream where to log audit events
+     * @param outputStreamOptions if {@code CLOSE} that should be closed in auditFinished()
+     * @throws IllegalArgumentException if outputStreamOptions is null
+     * @throws IOException if there is reading errors.
+     */
+    public SarifLogger(
+        OutputStream outputStream,
+        OutputStreamOptions outputStreamOptions) throws IOException {
+        writer = new PrintWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
+        if (outputStreamOptions == null) {
```

Comment:
please move this if condition as first statement, guard statements should go first

Reply:
Makes sense to be